File: 2501/solution.py
from math import sqrt


class Solution:
    def longestSquareStreak(self, nums: list[int]) -> int:
        numbers = set(nums)
        res = -1
        for num in nums:
            if sqrt(num) in numbers or num * num not in numbers:
                continue

            curr_count = 0
            curr = num
            while curr in numbers:
                curr_count += 1
                curr = curr * curr

            res = max(res, curr_count)
        return res

        # time O(n)
        # space O(n)

File: 2501/solution.java
class Solution {
    public int longestSquareStreak(int[] nums) {
        Set<Integer> uniques = new HashSet<>();
        for (int num : nums) {
            uniques.add(num);
        }
        int res = 1;
        for (int num : uniques) {
            int sqrt = (int) Math.sqrt(num);
            int streak = 1;
            if (sqrt * sqrt == num && uniques.contains(sqrt)) {
                continue;
            }
            while (num < 1000 && uniques.contains(num * num)) {
                num *= num;
                streak++;
            }
            res = Math.max(res, streak);
        }
        return res == 1 ? -1 : res;
    }
}
// time O(n)
// space O(n)